doc(unit-testing): Improvements, add example/explanation over how to use third-party resources and common issues (#1440)
operator-framework/operator-sdk: doc/user/unit-testing.md

<!-- Link vars declaration -->

> [doc_e2e_test]:./test-framework/writing-e2e-tests.md

Reviewer: Fix broken reference
```suggestion
[doc_e2e_test]:./../test-framework/writing-e2e-tests.md
```

Author: It did not pass in the tests as well. :-(

Author: @LiliC and @hasbro17 relative paths are not passing in the test so I keep it as it was.

## Using a Fake client

The `controller-runtime`'s fake client exposes the same set of operations as a typical client, but simply tracks objects rather than sending requests over a network. You can create a new fake client that tracks an initial set of objects with the following code:

```Go
import (
    "testing"
    cachev1alpha1 "github.com/example-inc/memcached-operator/pkg/apis/cache/v1alpha1"
    "k8s.io/apimachinery/pkg/runtime"
    "sigs.k8s.io/controller-runtime/pkg/client/fake"
)

func TestMemcachedController(t *testing.T) {
    ...
    // A Memcached object with metadata and spec.
    memcached := &cachev1alpha1.Memcached{
        ObjectMeta: metav1.ObjectMeta{
            Name:      "memcached",
            Namespace: "memcached-operator",
        },
    }

    // Objects to track in the fake client.
    objs := []runtime.Object{memcached}

    // Create a fake client to mock API calls.
    cl := fake.NewFakeClient(objs...)
    ...
}
```

> The `cl` will cache `memcached` in an internal object tracker so that CRUD operations via `cl` can be performed on it.

Reviewer 2: ```suggestion
The fake client `cl` will cache `memcached` in an internal object tracker so that CRUD operations via `cl` can be performed on it.
```

Author: 👍 done.

        TypeMeta: v1.TypeMeta{  // TODO (user): Remove the TypeMeta declared
            APIVersion: "v1",   // the correct value will be `"route.openshift.io/v1"`
            Kind:       "Route",
        },
        ObjectMeta: v1.ObjectMeta{
            Name:      name,
            Namespace: namespace,
            Labels:    ls,
        },
    }
    ...
```
Following another example of the issue that can be faced when the third-party resource schema was not added properly.  	

```shell
create a route: (no kind is registered for the type v1.Route in scheme "k8s.io/client-go/kubernetes/scheme/register.go:61")`
```

## How to increase the verbosity of the logs?

Following is a snippet code as an example to increase the verbosity of the logs in order to better troubleshoot your tests. 

```go
import (
    ...
    logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
    ...
)
func TestMemcachedController(t *testing.T) {	
    //dev logs
    logf.SetLogger(logf.ZapLogger(true))
    ...
}
```

<!-- Link vars declaration -->
<!-- NOTE: The CI has a bug and the test will not pass with _ . Use "-" instead of "_" -->
[doc-e2e-test]:./test-framework/writing-e2e-tests.md

> [doc-client]:./user/client.md

Reviewer 3: Can you try doing: 
```
[doc-client]:/doc/user/client.md
```

Author: Hi @LiliC,

Yes, I can. However, note that I tried with the full URL as it was before and the same error was faced. Also, I checked and all the lnks are working.

Author: Hi @LiliC,

Done. The same issue is faced. Note that I tried to keep the links as it was before as well and it did not work too.

Reviewer 3: @camilamacedo86 This is very weird indeed, so sorry about this! 

@AlexNPavel any ideas why this would be failing? When I check the link manually it works fine.

Author: Hi @LiliC,

Yes, it is. My guess that the CI just is checking the files which are changed and is not all docs which have vars for links. In this way, I think we could move forward and create a task to fix the CI.

Reviewer 4: Sorry for all the hassle with this markdown test. The `_` bug combined with the tester following very strict CommonMark formatting that most viewers ignore (just like most browsers ignore incorrect HTML and make assumptions when an error occurs) has been problematic. The issue here was actually very minor. CommonMark requires there to be an empty line before a reference link section (or, more specifically, a link reference definition cannot interrupt a paragraph: https://spec.commonmark.org/0.29/#example-182). However, this document had a comment before the references section, so it was not interpreted as valid reference links. I have pushed the fix for the links (as well as some minor formatting fixes that my editor automatically fixed) to the branch.

Author: Really tks @AlexNPavel. I was getting nuts with it :-)

# Unit testing
------------

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  <!-- *generated with [DocToc](https://github.com/thlorenz/doctoc)*  -->

- [Overview](#overview)
- [Using a Fake client](#using-a-fake-client)
- [Testing Reconcile](#testing-reconcile)
- [Testing with 3rd Party Resources](#testing-with-3rd-party-resources)
- [How to increase the verbosity of the logs?](#how-to-increase-the-verbosity-of-the-logs)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Overview

> Testing your operator should involve both unit and [e2e][doc-e2e-test] tests. Unit tests assess the expected outcomes of individual operator components without requiring coordination between components. Operator unit tests should test multiple scenarios likely to be encountered by your custom operator logic at runtime. Much of your custom logic will involve API server calls via a [client][doc-client]; `Reconcile()` in particular will be making API calls on each reconciliation loop. These API calls can be mocked by using `controller-runtime`'s [fake client][doc-cr-fake-client], perfect for unit testing. This document steps through writing a unit test for the [memcached-operator][repo-memcached-reconcile]'s `Reconcile()` method using a fake client.

Reviewer 3: I would revert back the change you made to e2e, "end-to-end" is clearer in my opinion as we are explaining in detail.

Author: Done.

Author: @AlexNPavel really sorry. I push this small change and lost yours over it. Could you please check it?

Reviewer 3: FYI: If you do `--force-with-lease` you can prevent things like these from happening in the future :)

Also pinging @joelanford @hasbro17 to re-review ;)

Author: Just to share the `--force-with-lease` also lose the pr contributions :-(
But I applied the contribution in  the original one 👍

            Namespace: namespace,
        },
    }
 
    res, err := r.Reconcile(req)
    if err != nil {
       t.Fatalf("reconcile: (%v)", err)
    }
    
    // Check the result of reconciliation to make sure it has the desired state.
    if !res.Requeue {
        t.Error("reconcile did not requeue request as expected")
    }
    
    // Check if deployment has been created and has the correct size.
    dep := &appsv1.Deployment{}
    err = r.cl.Get(context.TODO(), req.NamespacedName, dep)
    if err != nil {
        t.Fatalf("get deployment: (%v)", err)
    }
    
    // Check if the quantity of Replicas for this deployment is equals the specification
    dsize := *dep.Spec.Replicas
    if dsize != replicas {
        t.Errorf("dep size (%d) is not the expected size (%d)", dsize, replicas)
    }
}
```

**The above tests check if:**

- `Reconcile()` fails to find a Deployment object
- A Deployment is created
- The request is requeued in the expected manner
- The number of replicas in the created Deployment's spec is as expected.

**NOTE**: A unit test checking more cases can be found in our [`samples repo`][code_test_example].

## Testing with 3rd Party Resources

> You may have added third-party resources in your operator as described in the [`Advanced Topics section of the user guide`][user_guide]. In order to create a unit-test to test this kind of resources, it is required to add the Schema as it was done for your operator project.

Reviewer: @camilamacedo86 Just to be clear, you only need to pass in the updated scheme to your Reconciler if your handler is actually using it.
If say I don't have a `scheme *runtime.Scheme` in my Reconciler struct then I don't need to do this.
https://github.com/operator-framework/operator-sdk-samples/blob/master/memcached-operator/pkg/controller/memcached/memcached_controller.go#L74

Also:
```suggestion
You may have added third-party resources in your operator as described in the [`Advanced Topics section of the user guide`][user_guide]. In order to create a unit-test to test these kinds of resources, it might be necessary to update the Scheme with the third-party resources and pass it to your Reconciler.
```

Author: 👍

> Following is a snippet code as example using the [`v1.Route`][ocp_doc_v1_route] OpenShift Object Schema.

Reviewer: ```suggestion
The following code snippet is an example that adds the [`v1.Route`][ocp_doc_v1_route] OpenShift scheme to the ReconcileMemcached reconciler's scheme.
```

```go

import (
	...
	routev1 "github.com/openshift/api/route/v1"

> 	...

Reviewer: The tabs for the above lines are being shown as 8 spaces.
This should be 4 spaces like the rest of the doc.

Author: Good spot.

        ObjectMeta: v1.ObjectMeta{
            Name:      name,
            Namespace: namespace,
            Labels:    getAppLabels(name),
        },
    }
    
    s.AddKnownTypes(appv1alpha1.SchemeGroupVersion, memcached)
    
    // Create a fake client to mock API calls.
    cl := fake.NewFakeClient(objs...)
    
    // Create a ReconcileMemcached object with the scheme and fake client.
    r := &ReconcileMemcached{client: cl, scheme: s}
    ...
}
```

In this way, you will be able to get the mock object injected into the Reconcile as the following example.

```go
    route := &routev1.Route{}
    err = r.client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: namespace}, route)
    if err != nil {
        t.Fatalf("get route: (%v)", err)
    }
```

## Troubleshooting

### How to increase the verbosity of the logs?

Following is a snippet code as an example to increase the verbosity of the logs in order to better troubleshoot your tests.

```go
import (
    ...
    logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
    ...
)
func TestMemcachedController(t *testing.T) {	
    //dev logs
    logf.SetLogger(logf.ZapLogger(true))
    ...
}
```

> **NOTE** Recommended just for development purpose.

Reviewer: Nit: Probably remove this line since unit tests are run while in development.
This won't set the logging verbosity for the actual operator while it's running.
```suggestion
```

Author: HI @hasbro17 usually the unit tests are executed by the CI. But I agree that this note is not required. I think it will not a problem anyway too. Tks.

<!-- Link vars declaration -->

> [doc_e2e_test]:https://github.com/operator-framework/operator-sdk/blob/2f772d1dc2340dd19bdc3ec8c2dc9f0f77cc8297/doc/test-framework/writing-e2e-tests.md

Reviewer 3: Can you please you relative links: So, for example, this needs to be: 
```
[doc_e2e_test]:../test-framework/writing-e2e-tests.md
```

> [doc_client]:https://github.com/operator-framework/operator-sdk/blob/5c50126e7a112d67826894997eca143e12dc165f/doc/user/client.md

Reviewer 3: ```suggestion
[doc_client]:../user/client.md
```

[doc_reconcile]:https://godoc.org/sigs.k8s.io/controller-runtime/pkg/reconcile#Reconciler
[code_test_example]:https://github.com/operator-framework/operator-sdk-samples/blob/master/memcached-operator/pkg/controller/memcached/memcached_controller_test.go#L25
[user_guide]:https://github.com/operator-framework/operator-sdk/blob/c3a77eee021a1cc6a31861a0efd253dc23fbe591/doc/user-guide.md#register-with-the-managers-scheme
[ocp_doc_v1_route]:https://docs.openshift.com/container-platform/3.11/rest_api/apis-route.openshift.io/v1.Route.html

Unit testing
------------

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  <!-- *generated with [DocToc](https://github.com/thlorenz/doctoc)*  -->

- [Overview](#overview)
- [Using a Fake client](#using-a-fake-client)
- [Testing Reconcile](#testing-reconcile)
- [Testing with 3rd Party Resources](#testing-with-3rd-party-resources)
- [Troubleshooting](#troubleshooting)

>   - [How to increase the verbose of the logs?](#how-to-increase-the-verbose-of-the-logs)

Author: Done 👍

            Namespace: namespace,
        },
    }
 
    res, err := r.Reconcile(req)
    if err != nil {
       t.Fatalf("reconcile: (%v)", err)
    }
    
    // Check the result of reconciliation to make sure it has the desired state.
    if !res.Requeue {
        t.Error("reconcile did not requeue request as expected")
    }
    
    // Check if deployment has been created and has the correct size.
    dep := &appsv1.Deployment{}
    err = r.cl.Get(context.TODO(), req.NamespacedName, dep)
    if err != nil {
        t.Fatalf("get deployment: (%v)", err)
    }
    
    // Check if the quantity of Replicas for this deployment is equals the specification
    dsize := *dep.Spec.Replicas
    if dsize != replicas {
        t.Errorf("dep size (%d) is not the expected size (%d)", dsize, replicas)
    }
}
```

**The above tests check if:**

- `Reconcile()` fails to find a Deployment object
- A Deployment is created
- The request is requeued in the expected manner
- The number of replicas in the created Deployment's spec is as expected.

**NOTE**: A unit test checking more cases can be found in our [`samples repo`][code_test_example].

## Testing with 3rd Party Resources

> You may have added third-party resources in your operator as described in the [`Advanced Topics section of the user guide`][user_guide]. In order to create a unit-test to test this kind of resources is required to add the Schema as it was done for your operator project. 

Reviewer 3: ```suggestion
You may have added third-party resources in your operator as described in the [`Advanced Topics section of the user guide`][user_guide]. In order to create a unit-test to test this kind of resources, it is required to add the Schema as it was done for your operator project.
```

Author: Done 👍

func TestMemcachedController(t *testing.T) {	
    ... 
    // Register operator types with the runtime scheme.
    s := scheme.Scheme
    
    // Add route Openshift scheme
    if err := routev1.AddToScheme(s); err != nil {
        t.Fatalf("Unable to add route scheme: (%v)", err)
    }
    
    // Create the mock for the Route 
    route := &routev1.Route{
        ObjectMeta: v1.ObjectMeta{
            Name:      name,
            Namespace: namespace,
            Labels:    getAppLabels(name),
        },
    }
    
    s.AddKnownTypes(appv1alpha1.SchemeGroupVersion, memcached)
    
    // Create a fake client to mock API calls.
    cl := fake.NewFakeClient(objs...)
    
    // Create a ReconcileMemcached object with the scheme and fake client.
    r := &ReconcileMemcached{client: cl, scheme: s}
    ...
}
```

In this way, you will be able to get the mock object injected into the Reconcile as the following example.

```go
    route := &routev1.Route{}
    err = r.client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: namespace}, route)
    if err != nil {
        t.Fatalf("get route: (%v)", err)
    }
```

## Troubleshooting

> ### How to increase the verbose of the logs?

Author: 👍

> Following is a snippet code as example to increase the verbose of the logs in order to get troubleshooting your test. 

Reviewer 3: ```suggestion
Following is a snippet code as an example to increase the verbosity of the logs in order to better troubleshoot your tests.
```

Author: Done 👍

```go
import (
    ...
    logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
    ...
)
func TestMemcachedController(t *testing.T) {	
    //dev logs
    logf.SetLogger(logf.ZapLogger(true))
    ...
}
```
**NOTE** Recommended just for development purpose.

> ### Common known issues 

Author: H @LiliC,

It is fine. Please, just let me know your final decision and f I should or not remove it.

> ### Common known issues 

Reviewer 5: Yeah, we didn't really come to a final conclusion. I lean towards not having an FAQ section and instead trying to constantly update and reorganize documentation to make it easier to find answers to user questions in sections where those questions are relevant. I feel like FAQs have a tendency to become a dumping ground, which results in deteriorating quality of other documentation.

> ### Common known issues 

Author: Thank you @joelanford and @LiliC. The section was removed and the comments addressed. Please, let me know if you would like to address any other suggestion/change.

> ### Common known issues 

Author: PS.: relative paths for the link are not passing in the test so I keep it as it was. However, it still not passing :-(.
